proposals/lip-Introduce-sparse-Merkle-trees.md

```diff
@@ -0,0 +1,589 @@
+
+An inclusion proof is characterized by the following properties:
+
+*   `queryKeys`: An array of keys whose presence in the dataset has to be verified. 
+*   `proof`: The proof itself, consisting of 2 parts: `siblingHashes`, `queries`:
+    *   `siblingHashes`: An array of bytes necessary to compute the Merkle root. 
+    *   `queries`: An array of objects, one per key in `queryKeys`. Each object describes the inclusion proof of the corresponding key and contains the following properties:
+        *   `key`: A byte value indicating the presence or not of the corresponding key in the tree. 
+        *   `bitmap`: A byte value whose binary expansion indicates the presence of a branch or leaf node (digit 1) or an empty node (digit 0) in the path from the leaf node to the root (see Figure 2). 
+        *   `value`: A byte value giving the value of the node associated with `key.`
+
+The protocol for an inclusion proof is similar to the [one specified for regular Merkle trees](https://github.com/LiskHQ/lips/blob/master/proposals/lip-0031.md#proof-of-inclusion-protocol):
+
+*   Verifier: The Verifier knows `merkleRoot` and sends the Prover an array of keys, `queryKeys`, for which they wish to obtain a proof-of-inclusion;
+*   Prover:
+    *   The Prover finds the leaf node corresponding to each key `k` in `queryKeys` and assigns to each a `query` object with a key, a bitmap, and a value. The bitmap indicates the presence of default empty nodes along the path to the Merkle root. If the node exists, `query.key` is set to `k` and `query.value` to the node value; if the node does not exist, the Verifier returns an inclusion proof for another leaf or a default empty node in the path of the `k`. In this case, `query.key` and `query.value` are set to the leaf key and leaf value or to `k` and empty value, respectively.  
+    *   The Prover generates `siblingHashes` by merging together the sibling hashes of the single paths in the correct order of application, and transmits the complete proof to the Verifier. The complete proof contains a common array `siblingHashes` and an array `queries`, containing a query object for each key in `queryKeys`, respecting the same order.
+*   Verifier: The Verifier uses `siblingHashes` to recalculate the Merkle root starting from the leaf nodes derived from `queryKeys` and `queries`, and checks that it equals `merkleRoot`.
+
+As shown, a non-inclusion proof is generated by providing a proof of inclusion for another leaf or a default empty node in the path of the target leaf node, indicating that there are no other leaves in the subtree (see Figure 2). 
+
+## Specification
+
+For the rest of this proposal, we define the following constants:
+
+| Name          | Type    | Value       |
+| ------------- |---------| ------------|
+| EMPTY_HASH    | bytes   | SHA-256("") |
+| LEAF_PREFIX   | bytes   | 0x00        |
+| BRANCH_PREFIX | bytes   | 0x01        |
+| KEY_LENGTH    | integer | Set when initializing the data structure |
```

Comment:
Here an below it is not clear what unit is used for the key length.

Maybe the parameter name could be `KEY_LENGTH_BYTES` or something similar. An for the description here we could as "Length of the key in bytes."